pkg/scheduler/plugins/coescheduling/core/gang_cache.go

```diff
@@ -0,0 +1,232 @@
+limitations under the License.
+*/
+
+package core
+
+import (
+	"context"
+	"sync"
+
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	listerv1 "k8s.io/client-go/listers/core/v1"
+	"k8s.io/klog/v2"
+	"sigs.k8s.io/scheduler-plugins/pkg/apis/scheduling/v1alpha1"
+	pgclientset "sigs.k8s.io/scheduler-plugins/pkg/generated/clientset/versioned"
+
+	pglister "sigs.k8s.io/scheduler-plugins/pkg/generated/listers/scheduling/v1alpha1"
+
+	"github.com/koordinator-sh/koordinator/apis/extension"
+	"github.com/koordinator-sh/koordinator/apis/scheduling/config"
+)
+
+type GangCache struct {
+	lock       *sync.RWMutex
+	gangItems  map[string]*Gang
+	pluginArgs *config.CoschedulingArgs
+	podLister  listerv1.PodLister
+	pgLister   pglister.PodGroupLister
+	pgClient   pgclientset.Interface
+}
+
+func NewGangCache(args *config.CoschedulingArgs, podLister listerv1.PodLister, pgLister pglister.PodGroupLister, client pgclientset.Interface) *GangCache {
+	return &GangCache{
+		gangItems:  make(map[string]*Gang),
+		lock:       new(sync.RWMutex),
+		pluginArgs: args,
+		podLister:  podLister,
+		pgLister:   pgLister,
+		pgClient:   client,
+	}
+}
+
+func (gangCache *GangCache) createGangToCache(gangId string) *Gang {
+	gangCache.lock.Lock()
+	defer gangCache.lock.Unlock()
+
+	if gang, exist := gangCache.gangItems[gangId]; exist {
+		return gang
+	}
+
+	gang := NewGang(gangId)
+	gangCache.gangItems[gangId] = gang
+	return gang
+}
+
+func (gangCache *GangCache) getGangFromCache(gangId string) *Gang {
+	gangCache.lock.RLock()
+	defer gangCache.lock.RUnlock()
+
+	return gangCache.gangItems[gangId]
+}
+
+func (gangCache *GangCache) deleteGangFromCache(gangId string) {
+	gangCache.lock.Lock()
+	defer gangCache.lock.Unlock()
+
+	delete(gangCache.gangItems, gangId)
+}
+
+func (gangCache *GangCache) RecoverGangCache() {
+	// recover podGroup First
+	pgLister := gangCache.pgLister
+	pgList, err := pgLister.List(labels.NewSelector())
+	if err != nil {
+		klog.Errorf("RecoverGangCache podGroupList List error: %v", err.Error())
+	}
+	for _, podGroup := range pgList {
+		gangCache.onPodGroupAdd(podGroup)
+
+	}
+	// recover gang from pod
+	podLister := gangCache.podLister
+	podsList, err := podLister.List(labels.NewSelector())
+	if err != nil {
+		klog.Errorf("RecoverGangCache podsList List error: %v", err.Error())
+	}
+	for _, pod := range podsList {
+		gangCache.onPodAdd(pod)
+	}
+
+}
+
+func (gangCache *GangCache) onPodAdd(obj interface{}) {
+	pod, ok := obj.(*v1.Pod)
+	if !ok {
+		return
+	}
+
+	gangName := GetGangNameByPod(pod)
+	if gangName == "" {
+		return
+	}
+	var gang *Gang
+	// the gang is created in either CRD way or Annotation way
+	if _, exist := pod.Labels[v1alpha1.PodGroupLabel]; exist {
+		gangName := pod.Labels[v1alpha1.PodGroupLabel]
+		gangNamespace := pod.Namespace
+		gangId := GetNamespaceSplicingName(gangNamespace, gangName)
+		gang = gangCache.getGangFromCache(gangId)
```

Comment:
getGangFromCache add a createIfNotExist variable, if not exist, create directly. this is to make sure get and create under one lock protect, this may avoid multi-thread problem: A\B both get fail, then both create